File: include/extras-kguiaddons/libkiconutils.zig
const QtC = @import("qt6zig");
const qtc = @import("qt6c");
const qnamespace_enums = @import("../libqnamespace.zig").enums;
const std = @import("std");
pub const map_i32_qtcqicon = std.AutoHashMapUnmanaged(i32, QtC.QIcon);

/// https://api-staging.kde.org/kiconutils.html
pub const kiconutils = struct {
    /// [Qt documentation](https://api-staging.kde.org/kiconutils.html#addOverlay)
    ///
    /// ``` param1: QtC.QIcon, param2: QtC.QIcon, param3: qnamespace_enums.Corner ```
    pub fn AddOverlay(param1: ?*anyopaque, param2: ?*anyopaque, param3: i32) QtC.QIcon {
        return qtc.KIconUtils_AddOverlay(@ptrCast(param1), @ptrCast(param2), @intCast(param3));
    }

    /// [Qt documentation](https://api-staging.kde.org/kiconutils.html#addOverlays)
    ///
    /// ``` param1: QtC.QIcon, param2: map_i32_qtcqicon, allocator: std.mem.Allocator ```
    pub fn AddOverlays(param1: ?*anyopaque, param2: map_i32_qtcqicon, allocator: std.mem.Allocator) QtC.QIcon {
        const param2_keys = allocator.alloc(i32, param2.count()) catch @panic("kiconutils.AddOverlays: Memory allocation failed");
        defer allocator.free(param2_keys);
        const param2_values = allocator.alloc(QtC.QIcon, param2.count()) catch @panic("kiconutils.AddOverlays: Memory allocation failed");
        defer allocator.free(param2_values);
        var i: usize = 0;
        var param2_it = param2.iterator();
        while (param2_it.next()) |entry| {
            const key = entry.key_ptr.*;
            param2_keys[i] = @intCast(key);
            param2_values[i] = entry.value_ptr.*;
            i += 1;
        }
        const param2_map = qtc.libqt_map{
            .len = param2.count(),
            .keys = @ptrCast(param2_keys.ptr),
            .values = @ptrCast(param2_values.ptr),
        };
        return qtc.KIconUtils_AddOverlays(@ptrCast(param1), param2_map);
    }

    /// [Qt documentation](https://api-staging.kde.org/kiconutils.html#addOverlays)
    ///
    /// ``` param1: QtC.QIcon, param2: [][]const u8, allocator: std.mem.Allocator ```
    pub fn AddOverlays2(param1: ?*anyopaque, param2: [][]const u8, allocator: std.mem.Allocator) QtC.QIcon {
        var param2_arr = allocator.alloc(qtc.libqt_string, param2.len) catch @panic("kiconutils.AddOverlays2: Memory allocation failed");
        defer allocator.free(param2_arr);
        for (param2, 0..param2.len) |item, i| {
            param2_arr[i] = .{
                .len = item.len,
                .data = item.ptr,
            };
        }
        const param2_list = qtc.libqt_list{
            .len = param2.len,
            .data = param2_arr.ptr,
        };
        return qtc.KIconUtils_AddOverlays2(@ptrCast(param1), param2_list);
    }

    /// [Qt documentation](https://api-staging.kde.org/kiconutils.html#addOverlays)
    ///
    /// ``` param1: []const u8, param2: [][]const u8, allocator: std.mem.Allocator ```
    pub fn AddOverlays3(param1: []const u8, param2: [][]const u8, allocator: std.mem.Allocator) QtC.QIcon {
        const param1_str = qtc.libqt_string{
            .len = param1.len,
            .data = param1.ptr,
        };
        var param2_arr = allocator.alloc(qtc.libqt_string, param2.len) catch @panic("kiconutils.AddOverlays3: Memory allocation failed");
        defer allocator.free(param2_arr);
        for (param2, 0..param2.len) |item, i| {
            param2_arr[i] = .{
                .len = item.len,
                .data = item.ptr,
            };
        }
        const param2_list = qtc.libqt_list{
            .len = param2.len,
            .data = param2_arr.ptr,
        };
        return qtc.KIconUtils_AddOverlays3(param1_str, param2_list);
    }
};

File: src_original/extras-kguiaddons/libkiconutils.cpp
#include <KIconUtils>
#include <QHash>
#include <QIcon>
#include <QList>
#include <QString>
#include <QByteArray>
#include <cstring>
#include <kiconutils.h>
#include "libkiconutils.h"
#include "libkiconutils.hxx"

QIcon* KIconUtils_AddOverlay(const QIcon* param1, const QIcon* param2, int param3) {
    return new QIcon(KIconUtils::addOverlay(*param1, *param2, static_cast<Qt::Corner>(param3)));
}

QIcon* KIconUtils_AddOverlays(const QIcon* param1, const libqt_map /* of int to QIcon* */ param2) {
    QHash<Qt::Corner, QIcon> param2_QMap;
    param2_QMap.reserve(param2.len);
    int* param2_karr = static_cast<int*>(param2.keys);
    QIcon** param2_varr = static_cast<QIcon**>(param2.values);
    for (size_t i = 0; i < param2.len; ++i) {
        param2_QMap[static_cast<Qt::Corner>(param2_karr[i])] = *(param2_varr[i]);
    }
    return new QIcon(KIconUtils::addOverlays(*param1, param2_QMap));
}

QIcon* KIconUtils_AddOverlays2(const QIcon* param1, const libqt_list /* of libqt_string */ param2) {
    QList<QString> param2_QList;
    param2_QList.reserve(param2.len);
    libqt_string* param2_arr = static_cast<libqt_string*>(param2.data);
    for (size_t i = 0; i < param2.len; ++i) {
        QString param2_arr_i_QString = QString::fromUtf8(param2_arr[i].data, param2_arr[i].len);
        param2_QList.push_back(param2_arr_i_QString);
    }
    return new QIcon(KIconUtils::addOverlays(*param1, param2_QList));
}

QIcon* KIconUtils_AddOverlays3(const libqt_string param1, const libqt_list /* of libqt_string */ param2) {
    QString param1_QString = QString::fromUtf8(param1.data, param1.len);
    QList<QString> param2_QList;
    param2_QList.reserve(param2.len);
    libqt_string* param2_arr = static_cast<libqt_string*>(param2.data);
    for (size_t i = 0; i < param2.len; ++i) {
        QString param2_arr_i_QString = QString::fromUtf8(param2_arr[i].data, param2_arr[i].len);
        param2_QList.push_back(param2_arr_i_QString);
    }
    return new QIcon(KIconUtils::addOverlays(param1_QString, param2_QList));
}
